File: ext/examples/define_module.rs
use std::{
    io::{stderr, Write},
    sync::Arc,
};

use algebra::{
    module::FDModule, steenrod_evaluator::SteenrodEvaluator, AdemAlgebra, Algebra, GeneratedAlgebra,
};
use anyhow::anyhow;
use bivec::BiVec;
use fp::{
    prime::{Prime, ValidPrime},
    vector::FpVector,
};
use rustc_hash::FxHashMap as HashMap;
use serde_json::{json, Value};

pub fn get_gens() -> anyhow::Result<BiVec<Vec<String>>> {
    ext::utils::init_logging()?;

    // Query for generators
    eprintln!("Input generators. Press return to finish.");
    stderr().flush()?;

    let mut gens: BiVec<Vec<_>> = BiVec::new(0);
    loop {
        let gen_deg: Option<i32> = query::optional("Generator degree", str::parse);
        if gen_deg.is_none() {
            eprintln!("This is the list of generators and degrees:");
            for (i, deg_i_gens) in gens.iter_enum() {
                for gen in deg_i_gens.iter() {
                    eprint!("({i}, {gen}) ");
                }
            }
            eprintln!();
            if query::yes_no("Is it okay?") {
                break;
            } else {
                if query::yes_no("Start over?") {
                    gens = BiVec::new(0);
                }
                continue;
            }
        }

        let gen_deg = gen_deg.unwrap();

        gens.extend_negative(gen_deg, Vec::new());
        gens.extend_with(gen_deg, |_| Vec::new());

        let gen_name = query::with_default(
            "Generator name",
            &format!("x{gen_deg}{}", gens[gen_deg].len()).replace('-', "_"),
            |x| {
                match x.chars().next() {
                    Some(a) => {
                        if !a.is_alphabetic() {
                            return Err("variable name must start with a letter".to_string());
                        }
                    }
                    None => return Err("Variable name cannot be empty".to_string()),
                };
                for c in x.chars() {
                    if !c.is_alphanumeric() && c != '_' {
                        return Err(format!(
                            "Variable name cannot contain {c}. Should be alphanumeric and '_'"
                        ));
                    }
                }
                Ok(x.to_string())
            },
        );
        gens[gen_deg].push(gen_name);
    }
    Ok(gens)
}

pub fn gens_to_json(gens: &BiVec<Vec<String>>) -> serde_json::Value {
    let mut gens_json = json!({});
    for (i, deg_i_gens) in gens.iter_enum() {
        for gen in deg_i_gens {
            gens_json[gen] = json!(i);
        }
    }
    gens_json
}

pub fn interactive_module_define_fdmodule(
    output_json: &mut Value,
    p: ValidPrime,
) -> anyhow::Result<()> {
    output_json["p"] = Value::from(p.as_u32());
    let algebra = Arc::new(AdemAlgebra::new(p, false));

    let gens = get_gens()?;
    let min_degree = gens.min_degree();
    let max_degree = gens.len();

    algebra.compute_basis(max_degree - min_degree);

    let mut graded_dim = BiVec::with_capacity(min_degree, max_degree);
    for i in gens.iter().map(Vec::len) {
        graded_dim.push(i);
    }

    let mut module = FDModule::new(Arc::clone(&algebra), String::new(), graded_dim);

    for (i, deg_i_gens) in gens.iter_enum() {
        for (j, gen) in deg_i_gens.iter().enumerate() {
            module.set_basis_element_name(i, j, gen.to_string());
        }
    }

    eprintln!(
        "Input actions. Write the value of the action in the form 'a x0 + b x1 + ...' where a, b \
         are non-negative integers and x0, x1 are names of the generators. The coefficient can be \
         omitted if it is 1"
    );

    let len = gens.len();
    for input_deg in gens.range().rev() {
        for output_deg in (input_deg + 1)..len {
            let op_deg = output_deg - input_deg;
            if gens[output_deg].is_empty() {
                continue;
            }
            for op_idx in algebra.generators(op_deg) {
                for input_idx in 0..gens[input_deg].len() {
                    let output = query::raw(
                        &format!(
                            "{} {}",
                            algebra.basis_element_to_string(op_deg, op_idx),
                            gens[input_deg][input_idx]
                        ),
                        |expr| {
                            let mut result = vec![0; gens[output_deg].len()];
                            if expr == "0" {
                                return Ok(result);
                            }
                            for term in expr.split('+') {
                                let term = term.trim();
                                let (coef, gen) = match term.split_once(' ') {
                                    Some((coef, gen)) => (str::parse::<u32>(coef)?, gen),
                                    None => (1, term),
                                };

                                if let Some(gen_idx) =
                                    gens[output_deg].iter().position(|d| d == gen)
                                {
                                    result[gen_idx] += coef;
                                } else {
                                    return Err(anyhow!(
                                        "No generator {gen} in degree {output_deg}"
                                    ));
                                }
                            }

                            Ok(result)
                        },
                    );

                    module.set_action(op_deg, op_idx, input_deg, input_idx, &output);
                }
            }
            module.extend_actions(input_deg, output_deg);
            module.check_validity(input_deg, output_deg)?;
        }
    }

    module.to_json(output_json);
    Ok(())
}

/// Given a string representation of an element in an algebra together with a generator, multiply
/// each term on the right with the generator.
fn replace(algebra_elt: &str, gen: &str) -> String {
    algebra_elt.replace('+', &format!("{gen} +")) + " " + gen
}

pub fn interactive_module_define_fpmodule(
    output_json: &mut Value,
    p: ValidPrime,
) -> anyhow::Result<()> {
    let gens = get_gens()?;
    let min_degree = gens.min_degree();
    let max_degree = gens.len();

    let ev = SteenrodEvaluator::new(p);

    let mut graded_dim = BiVec::with_capacity(min_degree, max_degree);
    for i in gens.iter().map(Vec::len) {
        graded_dim.push(i);
    }

    eprintln!("Input relations");
    match p.as_u32() {
        2 => eprintln!("Write relations in the form 'Sq6 * Sq2 * x + Sq7 * y'"),
        _ => eprintln!(
            "Write relations in the form 'Q5 * P(5) * x + 2 * P(1, 3) * Q2 * y', where P(...) and \
             Qi are Milnor basis elements."
        ),
    }

    let mut degree_lookup = HashMap::default();
    for (i, deg_i_gens) in gens.iter_enum() {
        for gen in deg_i_gens.iter() {
            degree_lookup.insert(gen.clone(), i);
        }
    }

    let mut adem_relations = Vec::new();
    let mut milnor_relations = Vec::new();

    loop {
        let relation = query::raw("Enter relation", |rel| {
            let result = ev.evaluate_module_adem(rel)?;

            if result.is_empty() {
                return Ok(result);
            }

            // Check that the generators exist and the terms all have the same degree
            let mut deg = None;
            for (gen, (op_deg, _)) in result.iter() {
                let cur_deg = op_deg
                    + degree_lookup
                        .get(gen)
                        .ok_or_else(|| anyhow!("Unknown generator: {gen}"))?;
                if deg.is_none() {
                    deg = Some(cur_deg);
                } else if deg != Some(cur_deg) {
                    return Err(anyhow!(
                        "Relation terms have different degrees: {} and {cur_deg}",
                        deg.unwrap()
                    ));
                }
            }

            Ok(result)
        });

        if relation.is_empty() {
            break;
        }

        let mut adem_relation = String::new();
        let mut milnor_relation = String::new();

        let mut milnor_op = FpVector::new(p, 0);
        for (gen, (op_deg, adem_op)) in relation.iter() {
            if adem_op.is_zero() {
                continue;
            }
            if !adem_relation.is_empty() {
                adem_relation += " + ";
                milnor_relation += " + ";
            }
            milnor_op.set_scratch_vector_size(adem_op.len());
            ev.adem_to_milnor(&mut milnor_op, 1, *op_deg, adem_op);

            adem_relation += &replace(&ev.adem.element_to_string(*op_deg, adem_op.as_slice()), gen);
            milnor_relation += &replace(
                &ev.milnor.element_to_string(*op_deg, milnor_op.as_slice()),
                gen,
            );
        }
        if !adem_relation.is_empty() {
            adem_relations.push(Value::String(adem_relation));
            milnor_relations.push(Value::String(milnor_relation));
        }
    }

    output_json["p"] = Value::from(p.as_u32());
    output_json["type"] = Value::String("finitely presented module".to_owned());
    for (i, deg_i_gens) in gens.iter_enum() {
        for gen in deg_i_gens {
            output_json["gens"][gen] = Value::from(i);
        }
    }
    output_json["adem_relations"] = Value::Array(adem_relations);
    output_json["milnor_relations"] = Value::Array(milnor_relations);
    Ok(())
}

fn main() -> anyhow::Result<()> {
    let module_type = query::with_default(
        "Input module type (default 'finite dimensional module'):\n (fd) - finite dimensional \
         module \n (fp) - finitely presented module\n",
        "fd",
        |x| match x {
            "fd" | "fp" => Ok(x.to_string()),
            _ => Err(format!("Invalid type '{x}'. Type must be 'fd' or 'fp'")),
        },
    );

    let p: ValidPrime = query::with_default("p", "2", str::parse);
    let mut output_json = json!({});

    eprintln!("module_type: {module_type}");
    match &*module_type {
        "fd" => interactive_module_define_fdmodule(&mut output_json, p)?,
        "fp" => interactive_module_define_fpmodule(&mut output_json, p)?,
        _ => unreachable!(),
    }
    println!("{output_json}");
    Ok(())
}

File: ext_py/examples/define_module.py
#!/usr/bin/env python3
"""
Interactive module definition tool for creating finite dimensional or finitely presented modules.
Python translation of define_module.rs example.
"""

import ext
import json
from typing import Dict, List, Any


def get_generators() -> Dict[int, List[str]]:
    """Interactively get generators from user."""
    print("Input generators. Press return to finish.")

    generators = {}

    while True:
        degree_input = input("Generator degree (or press Enter to finish): ").strip()
        if not degree_input:
            if generators:
                print("Generators and degrees:")
                for deg, gens in generators.items():
                    for gen in gens:
                        print(f"({deg}, {gen})", end=" ")
                print()

                if input("Is it okay? (y/n): ").lower().startswith("y"):
                    break
                else:
                    if input("Start over? (y/n): ").lower().startswith("y"):
                        generators = {}
                    continue
            else:
                break

        try:
            degree = int(degree_input)
        except ValueError:
            print("Invalid degree. Please enter an integer.")
            continue

        if degree not in generators:
            generators[degree] = []

        default_name = f"x{degree}{len(generators[degree])}".replace("-", "_")
        gen_name = (
            input(f"Generator name (default '{default_name}'): ").strip()
            or default_name
        )

        # Validate generator name
        if not gen_name[0].isalpha():
            print("Variable name must start with a letter")
            continue
        if not all(c.isalnum() or c == "_" for c in gen_name):
            print("Variable name must be alphanumeric with underscores only")
            continue

        generators[degree].append(gen_name)

    return generators


def define_finite_dimensional_module(prime: int) -> Dict[str, Any]:
    """Define a finite dimensional module interactively."""
    output = {"p": prime, "type": "finite dimensional module"}

    # Create algebra
    alg = ext.AdemAlgebra(prime=prime, truncated=False)

    # Get generators
    generators = get_generators()
    min_degree = min(generators.keys()) if generators else 0
    max_degree = max(generators.keys()) if generators else 0

    # Compute algebra basis
    alg.compute_basis(max_degree - min_degree)

    # Create module
    graded_dims = [
        len(generators.get(i, [])) for i in range(min_degree, max_degree + 1)
    ]
    module = ext.FDModuleBuilder(alg, "", graded_dims, min_degree)

    # Set generator names
    for degree, gen_list in generators.items():
        for idx, gen_name in enumerate(gen_list):
            module.set_basis_element_name(degree, idx, gen_name)

    # Input actions
    print(
        "Input actions. Write the value of the action in the form 'a x0 + b x1 + ...'"
    )
    print("where a, b are non-negative integers and x0, x1 are generator names.")
    print("The coefficient can be omitted if it is 1.")

    for input_deg in sorted(generators.keys(), reverse=True):
        for output_deg in range(input_deg + 1, max_degree + 1):
            if output_deg not in generators or not generators[output_deg]:
                continue

            op_deg = output_deg - input_deg

            for op_idx in alg.generators(op_deg):
                for input_idx, input_gen in enumerate(generators[input_deg]):
                    op_string = alg.basis_element_to_string(op_deg, op_idx)
                    prompt = f"{op_string} {input_gen}: "

                    while True:
                        expression = input(prompt).strip()
                        try:
                            result = parse_linear_combination(
                                expression, generators[output_deg]
                            )
                            module.set_action(
                                op_deg, op_idx, input_deg, input_idx, result
                            )
                            break
                        except ValueError as e:
                            print(f"Error: {e}")

            module.extend_actions(input_deg, output_deg)
            module.check_validity(input_deg, output_deg)

    # Convert to JSON
    output.update(module.to_json())
    return output


def define_finitely_presented_module(prime: int) -> Dict[str, Any]:
    """Define a finitely presented module interactively."""
    output = {"p": prime, "type": "finitely presented module"}

    # Get generators
    generators = get_generators()

    # Create evaluator
    evaluator = ext.SteenrodEvaluator(prime)

    # Set up generator lookup
    generator_degrees = {}
    for degree, gen_list in generators.items():
        for gen in gen_list:
            generator_degrees[gen] = degree

    # Input relations
    print("Input relations")
    if prime == 2:
        print("Write relations in the form 'Sq6 * Sq2 * x + Sq7 * y'")
    else:
        print("Write relations in the form 'Q5 * P(5) * x + 2 * P(1, 3) * Q2 * y'")
        print("where P(...) and Qi are Milnor basis elements.")

    adem_relations = []
    milnor_relations = []

    while True:
        relation = input("Enter relation (or press Enter to finish): ").strip()
        if not relation:
            break

        try:
            # Parse and validate relation
            result = evaluator.evaluate_module_adem(relation)

            if not result:
                continue

            # Check degrees are consistent
            degrees = []
            for gen, (op_deg, _) in result.items():
                total_deg = op_deg + generator_degrees[gen]
                degrees.append(total_deg)

            if len(set(degrees)) > 1:
                print(f"Error: Relation terms have different degrees: {degrees}")
                continue

            # Convert to string representations
            adem_terms = []
            milnor_terms = []

            for gen, (op_deg, adem_op) in result.items():
                if not adem_op.is_zero():
                    adem_str = evaluator.adem_element_to_string(op_deg, adem_op)
                    milnor_op = evaluator.adem_to_milnor(op_deg, adem_op)
                    milnor_str = evaluator.milnor_element_to_string(op_deg, milnor_op)

                    adem_terms.append(f"{adem_str} {gen}")
                    milnor_terms.append(f"{milnor_str} {gen}")

            if adem_terms:
                adem_relations.append(" + ".join(adem_terms))
                milnor_relations.append(" + ".join(milnor_terms))

        except Exception as e:
            print(f"Error parsing relation: {e}")

    # Build output JSON
    output["gens"] = {
        gen: deg for deg, gen_list in generators.items() for gen in gen_list
    }
    output["adem_relations"] = adem_relations
    output["milnor_relations"] = milnor_relations

    return output


def parse_linear_combination(expression: str, generator_names: List[str]) -> List[int]:
    """Parse a linear combination like 'a x0 + b x1' into coefficients."""
    if expression == "0":
        return [0] * len(generator_names)

    result = [0] * len(generator_names)

    for term in expression.split("+"):
        term = term.strip()
        if " " in term:
            coef_str, gen = term.split(" ", 1)
            coef = int(coef_str)
        else:
            coef = 1
            gen = term

        try:
            gen_idx = generator_names.index(gen)
            result[gen_idx] += coef
        except ValueError:
            raise ValueError(f"Unknown generator: {gen}")

    return result


def main():
    # Get module type
    print("Input module type:")
    print("(fd) - finite dimensional module")
    print("(fp) - finitely presented module")
    module_type = input("Type (default 'fd'): ").strip() or "fd"

    if module_type not in ["fd", "fp"]:
        print(f"Invalid type '{module_type}'. Type must be 'fd' or 'fp'")
        return

    # Get prime
    prime = int(input("p (default 2): ").strip() or "2")

    print(f"module_type: {module_type}")

    if module_type == "fd":
        output = define_finite_dimensional_module(prime)
    else:
        output = define_finitely_presented_module(prime)

    print(json.dumps(output, indent=2))


if __name__ == "__main__":
    main()
